akka-actor-typed/src/main/scala/akka/actor/typed/RoutingHashExtractor.scala

```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.actor.typed
+
+/**
+ * Used in [[akka.actor.typed.javadsl.Routers]] and [[akka.actor.typed.scaladsl.Routers]].
+ *
+ * Implementations of this class will be used in consistent hashing process. Result of this operation
+ * should possibly uniquely distinguish messages.
+ *
+ * If you have String based key, you can easily transform it to Array[Byte] using either
+ * [[akka.actor.typed.javadsl.RoutingHashExtractor]] or [[akka.actor.typed.scaladsl.RoutingHashExtractor]]
+ */
+@FunctionalInterface
+trait RoutingHashExtractor[T] {
+
+  def apply(message: T): Array[Byte]
```

Comment:
In many cases I think it will be more natural to use a String as the hash key. One question is if we should use `String` instead of `Byte`?

Reply:
Yeah we discussed this back and forth a bit, last idea was to have this as the extractor but provide a factory in the companion object for the `T => String` case.

Reply:
but that is more difficult to discover if String is the normal case, and it would be some overhead to convert to Array[Byte] from String. Also, if not a plain String it's probably the concatenation of a few fields in the message, and converting those first to Array[Byte] isn't very obvious.